lib/queue.js

```diff
@@ -603,14 +642,34 @@ Queue.prototype.startMoveUnlockedJobsToWait = function() {
   }
 };
 
-Queue.prototype.processJobs = function(resolve, reject){
-  var _this = this;
-  var processJobs = this.processJobs.bind(this, resolve, reject);
 
+Queue.prototype.processJobs = function(resolve, reject, options){
+  var _this = this;
+  var processJobs = this.processJobs.bind(this, resolve, reject, options);
   if(!this.closing){
+    var idleTime = 0;
+    var obeyRateLimit = function() {};
+
+    if (options.rateLimiter) {
+      obeyRateLimit = function() {
+        return options.rateLimiter()
+          .then(function(timeLeft) {
+            if (!timeLeft) { return; }
+            idleTime += timeLeft;
+            return Promise.delay(timeLeft)
+              .then(obeyRateLimit);
+          });
```

Comment:
I am a bit unsure, but I think this kind of asynchronous recursive structure will leak memory. We had to use a process.nextTick in order to "break" the promise chain and remove a leak we had in processJobs